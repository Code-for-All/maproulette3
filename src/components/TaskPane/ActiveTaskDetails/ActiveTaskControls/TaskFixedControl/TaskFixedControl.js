import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { FormattedMessage } from 'react-intl'
import classNames from 'classnames'
import { TaskStatus }
       from '../../../../../services/Task/TaskStatus/TaskStatus'
import messages from './Messages'

/**
 * TaskFixedControl displays a control for marking a task with a fixed status.
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export default class TaskFixedControl extends Component {
  render() {
    return (
      <button className={classNames("button large-and-wide full-width label-only fixed-control",
                                    this.props.className)}
              onClick={() => this.props.complete(TaskStatus.fixed)}>
        <FormattedMessage {...messages.fixedLabel} />
      </button>
    )
  }
}

TaskFixedControl.propTypes = {
  /** Invoked to mark the task as already-fixed */
  complete: PropTypes.func.isRequired,
}
