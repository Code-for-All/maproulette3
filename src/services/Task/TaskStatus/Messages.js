import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with TaskStatus.
 */
export default defineMessages({
  created: {
    id: "Task.status.created",
    defaultMessage: "Created"
  },
  fixed: {
    id: "Task.status.fixed",
    defaultMessage: "Fixed"
  },
  falsePositive: {
    id: "Task.status.falsePositive",
    defaultMessage: "Not an Issue"
  },
  skipped: {
    id: "Task.status.skipped",
    defaultMessage: "Skipped"
  },
  deleted: {
    id: "Task.status.deleted",
    defaultMessage: "Deleted"
  },
  alreadyFixed: {
    id: "Task.status.alreadyFixed",
    defaultMessage: "Already Fixed"
  },
  tooHard: {
    id: "Task.status.tooHard",
    defaultMessage: "Too Hard"
  },
})
