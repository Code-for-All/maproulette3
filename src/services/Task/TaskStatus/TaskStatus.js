import { map as _map,
         invert as _invert,
         fromPairs as _fromPairs } from 'lodash'
import messages from './Messages'

// These statuses are defined on the server
export const TASK_STATUS_CREATED = 0
export const TASK_STATUS_FIXED = 1
export const TASK_STATUS_FALSE_POSITIVE = 2
export const TASK_STATUS_SKIPPED = 3
export const TASK_STATUS_DELETED = 4
export const TASK_STATUS_ALREADY_FIXED = 5
export const TASK_STATUS_TOO_HARD = 6

export const TaskStatus = Object.freeze({
  created: TASK_STATUS_CREATED,
  fixed: TASK_STATUS_FIXED,
  falsePositive: TASK_STATUS_FALSE_POSITIVE,
  skipped: TASK_STATUS_SKIPPED,
  deleted: TASK_STATUS_DELETED,
  alreadyFixed: TASK_STATUS_ALREADY_FIXED,
  tooHard: TASK_STATUS_TOO_HARD,
})

export const keysByStatus = Object.freeze(_invert(TaskStatus))

/**
 * Returns true if the given status represents completion of
 * the task.
 */
export const isCompleted = function(status) {
  return status === TaskStatus.fixed ||
         status === TaskStatus.falsePositive ||
         status === TaskStatus.alreadyFixed ||
         status === TaskStatus.deleted
}

/**
 * Returns an object mapping difficulty values to raw internationalized
 * messages suitable for use with FormattedMessage or formatMessage.
 */
export const messagesByStatus = _fromPairs(
  _map(messages, (message, key) => [TaskStatus[key], message])
)

/** Returns object containing localized labels  */
export const statusLabels = intl => _fromPairs(
  _map(messages, (message, key) => [key, intl.formatMessage(message)])
)
