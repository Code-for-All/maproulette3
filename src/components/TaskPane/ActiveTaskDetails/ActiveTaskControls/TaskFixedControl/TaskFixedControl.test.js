import React from 'react'
import TaskFixedControl from './TaskFixedControl'
import { TaskStatus } from '../../../../../services/Task/TaskStatus/TaskStatus'

let basicProps = null

beforeEach(() => {
  basicProps = {
    task: {
      id: 123,
      parent: {
        id: 321,
      },
      status: TaskStatus.created,
    },
    complete: jest.fn(),
    intl: {formatMessage: jest.fn(m => m.defaultMessage)},
  }
})

test("shows fixed control", () => {
  const wrapper = shallow(
    <TaskFixedControl {...basicProps} />
  )

  expect(wrapper.find('.fixed-control').exists()).toBe(true)

  expect(wrapper).toMatchSnapshot()
})

test("clicking the fixed button signals task completion with fixed status", () => {
  const wrapper = shallow(
    <TaskFixedControl {...basicProps} />
  )

  wrapper.find('.fixed-control').simulate('click')

  expect(basicProps.complete).toBeCalledWith(TaskStatus.fixed)
})
