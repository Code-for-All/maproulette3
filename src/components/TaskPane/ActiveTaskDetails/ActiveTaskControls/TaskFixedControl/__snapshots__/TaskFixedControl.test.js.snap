// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows fixed control 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskFixedControl
    complete={[Function]}
    intl={
        Object {
            "formatMessage": [Function],
          }
    }
    task={
        Object {
            "id": 123,
            "parent": Object {
              "id": 321,
            },
            "status": 0,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <FormattedMessage
        defaultMessage="I fixed it!"
        id="Task.controls.fixed.label"
        values={Object {}}
/>,
      "className": "button large-and-wide full-width label-only fixed-control",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "defaultMessage": "I fixed it!",
        "id": "Task.controls.fixed.label",
        "values": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <FormattedMessage
          defaultMessage="I fixed it!"
          id="Task.controls.fixed.label"
          values={Object {}}
/>,
        "className": "button large-and-wide full-width label-only fixed-control",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "defaultMessage": "I fixed it!",
          "id": "Task.controls.fixed.label",
          "values": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
